_posts/2024-12-26-issues-with-microros.md

```diff
@@ -0,0 +1,230 @@
+6. Next blocking delay occurs, repeating the cycle
+
+### Leaky Abstraction
+A leaky abstraction occurs when high-level interface (microros in our case) fails to completely hide low-level implementation details, forcing you to understand the underlying complexity that it was trying to hide. My initial reasoning was that we should avoid communication and use an abstraction. However, the "simple" abstraction provided by micro-ROS forced us to confront complex , including:
+
+- Network transport layers
+- Message buffering mechanisms
+
+In hindsight, there is a couple of solutions we could have implemented, but we did not understand the system at the time.
+We ended up using a simpler solution: UART communication. 
+
+### A Simpler Solution
+this an example of sending motor commands
+
+```cpp
+void tx_task() {
+    char send_buf[16];
+    while (true) {
+        send_buf[0] = 's';  // start marker
+        send_buf[1] = 'f';  // data type
+        memcpy(send_buf + 3, &imu_data.roll, sizeof(float));
+        uart_write_bytes(UART_NUM_1, send_buf, sizeof(send_buf));
+        vTaskDelay(50 / portTICK_PERIOD_MS);
+    }
+}
+
+```
+
+THAT IT! 
+Other than the serialization and deserialization, the system was:
+
+- Predictable
+- Easy to debug
+- Free from hidden complexities
+- Direct and transparent
+
+## The Paradox of Abstration
+
+Here's the paradox of abstractions that Joel Spolsky points out in his blog: 
+> "Abstractions save us time working, but they don't save us time learning."
+
+When abstractions work, they're magical! But when they fail, you need to understand:
+
+1. The abstraction itself
+2. The underlying system it's hiding
+3. How they interact
+
+## How to Make the Choice
+So, should you use micro-ROS in your project?
+
+As with any engineering decisions, there's unfortunately no one-size-fits-all answer. 
+
+However, here’s a framework that may guide your decision making, not just for micro-ROS, but for any tool:
+
+Before adopting any new technology, ask yourself 2 questions:
+ - What's the learning budget?
+ - What's the complexity-benefit ratio
+
+
+Some frameworks, are worth the investment because they solve a problem harder than. 
+For example, we stuck with ROS for our newton project because of the perception and planning aspect introduces camera, imu and sensor fusion that was easier to deal with ROS
+Here's the topic list
+We have encoured a lot of annoyance with the framework but the problem solved is worth. 
+
+However, for other projects, like our TWIP controller, the learning investment might outweigh the benefits.
+You may not know whe the complexity-benefit ratio is, but keep a mental note as you go. 
+
+
+## The Takeaway 
+
+Whether you choose micro-ROS or any other tools, be prepared for the leaks. 
+In the world of software, they're not a matter of if, but when. Sometimes, as we learned, the simplest solution might be the most robust one.
```

Comment:
```suggestion
In the world of software, they're not a matter of if, but when. Sometimes, as we learned, the simplest solution might be the most robust one. Or it might be too robust, but requires heavy investment in learning it.
```